### ISSUE TITLE
unable to download specific artists works (may be on others too)

### ISSUE THREAD

**Issue by driver240b** [Simon-Tesla/RaccoonyWebEx#90]
Went to download an artist's gallery and got a download error. Raccoony seemed unable to download the files as they showed up for the artist (https://www.furaffinity.net/user/angelina~marie/). Settings were not changed from previous artists, so it is unknown why this suddenly happened here.

**To Reproduce**
1) go to https://www.furaffinity.net/gallery/angelina~marie/
2) open any tab(s) for download. auto-download optional, I had it on.
3) attempt to download any file(s)

**Expected behavior**
You should see it attempt to download, then change to "download error"

**Screenshots**
*NOTE: the following screenshots contain NSFW images.*

This displays error
[error 1](https://user-images.githubusercontent.com/80655131/111090404-9c5f0f80-84fd-11eb-9e79-e8524bd2de68.png)

This one shows the artist page and settings
[error 3](https://user-images.githubusercontent.com/80655131/111090448-c1538280-84fd-11eb-878c-3ff7d1b83913.png)

These two show the downloading of my recent submissions. Note how it is downloading them with no issues
[error 4](https://user-images.githubusercontent.com/80655131/111090494-ed6f0380-84fd-11eb-98ee-cd455caa2b6f.png)
[error 5](https://user-images.githubusercontent.com/80655131/111090501-eea03080-84fd-11eb-88ed-9febf78effe2.png)

**Environment**
 - Windows 10 pro
 - Chrome ver. 89.0.4389.82
 - Raccoony 1.2.4

**Additional context**
 - https://www.furaffinity.net/gallery/angelina~marie/
 - Settings:
[raccoony-settings.txt](https://github.com/Simon-Tesla/RaccoonyWebEx/files/6138044/raccoony-settings.txt)
 - This may occur on other artist's pages. This is just the first time I have seen it happen, so I mention this one in particular. May revise if other artists are found to have this bug.

**Comment by Simon-Tesla** [comment]
Did a quick check, doesn't appear to repro in Firefox, but does repro in Chrome.

(Also switched the images to links since they contained NSFW imagery.)

**Comment by Simon-Tesla** [comment]
Chrome throws an "invalid filename" exception when attempting to create the file; I suspect it's something to do with the '~' in the artist's username.

**Comment by Eupeptic1** [comment]
I find the same thing on Linux: does not repro on Firefox (78.8.0esr on Debian 9), but does repro on Chrome and Edge (89.0.4389.90 and 91.0.825.0 dev, respectively, both on Debian 10).  Chrome and Edge never even get to the point of making the directory for the artist in `~/Download/raccoony`.

A little Googling suggests that `~` is not actually an illegal character in a file or directory name in FAT(12|16|32) or NTFS.  I'm almost 100% sure it's not an illegal character in a file or directory name on ext2/3/4 on Linux; I tried it!

[This](https://stackoverflow.com/questions/29294910/unable-to-add-files-with-name-containing-tilde-followed-by-a-number) is for a different Windows application, but in their case, the application disallows tildes in filenames because of a claimed possible interaction with how NTFS generates 8.3 filenames.

Aha!  Google cites the same problem.  The [bug report (from 2015)](https://bugs.chromium.org/p/chromium/issues/detail?id=479419) says it's WONTFIX and the [changelog](https://chromium.googlesource.com/chromium/src/+/92a354fca8d118ad1be38612f85aa71a9a25303b) has a little more information on the characters they banned.  Apparently this happened at or near Chrome 42.

I guess the questions are:

1) What, if anything, should Raccoony do about this?

2) If Raccoony does something, should it do the same thing for every browser, or behave differently for Firefox vs Chrome-and-friends?

3) If Raccoony does something, should it happen in each site plugin (seems kinda redundant) or at the next higher level?

As a quickie test, I added this in `furaffinity.ts`, right before I.Media gets filled in

    // FIXME: experimental tilde removal
    let notildes = username.replace("~", "_");
    logger.log("fa: no-tilde username ", notildes);
    
    const result: I.Media = {
      url: url,
      siteName: serviceName,
      submissionId: id,
      previewUrl: previewUrl,
      author: notildes,

and it then works in Chrome on Linux; you get a directory of `~/Downloads/raccoony/angelina_marie`, and the filename has `_by_angelina_marie` in it.  I have not tested more extensively than that.

**Comment by Simon-Tesla** [comment]
@Eupeptic1 I already have code to replace characters that aren't allowed in filenames in the sanitizePath function in download.ts, so whatever we do should be added there. Either we could add it to the existing regex (which would change the download path for FF users that have previously downloaded something with a ~), or we could add another check to see if we're on Chrome (or probably more realistically, not FF) and do the substitution only then.

**Comment by Eupeptic1** [comment]
It turns out that [TMTOWTDI](https://stackoverflow.com/questions/61070761/how-can-i-know-if-browser-is-chrome-vs-firefox-from-web-extension-popup-javascri), which is good, because some of them don't work.  :D

It also turns out that a clever mouse already had the working test coded into the top of `download.ts`, possibly for future use.

I put `furaffinity.ts` back to stock, and then I added this to `sanitizePath()`, right before "Replace any significant OS characters..."

    // FIXME Chrome bans tildes in file and dir names, so remove them
    if(!isFirefox) {
      pathPart = pathPart.replace(/~/g, '_');
      }

I tested this on one image from angelina~marie.  I then dug around a bit and found a couple of other artists on FA that have tildes in their usernames, and tested it with three images from each of these two artists.  Note that **these links may contain both SFW and NSFW images**, so they have been munged.

https://www.furaffinity dot net/gallery/\~anderwix\~/
https://www.furaffinity dot net/gallery/ashurrr~/

Results:

Chrome and Edge on Linux give directories of `angelina_marie`, `_anderwix_`, and `ashurrr_`, and filenames with `_by_angelina_marie`, `_by__anderwix_`, and `_by_ashurrr_`.

Firefox on Linux gives directories of `angelina~marie`, `~anderwix~`, and `ashurrr~`, and filenames with `_by_angelina~marie`, `_by_~anderwix~`, and `_by_ashurrr~`.

The image files and metadata files from Chrome, Edge, and Firefox all compare binary equal.

From all three browsers, the metadata file has human Author: and JSON author with the artist's name with the tilde(s) in it.

From all three browsers, the metadata file has JSON url with the artist's name with the tilde(s) in it.

(I'm testing on Edge because 1) I already have it set up and 2) it's technically Chromium and not Chrome, so I thought it might be an additional way to make sure the "not Firefox" check works.)
